File: app/javascript/channels/chat_channel.js
import consumer from "./consumer"

document.addEventListener('turbo:load', () => {
  const roomElement = document.getElementById('room-id');
  if (roomElement) {
    const roomId = roomElement.getAttribute('data-room-id');

    consumer.subscriptions.create({ channel: "ChatChannel", room: roomId }, {
      connected() {
        console.log("Connected to the chat channel.");
      },

      disconnected() {
        console.log("Disconnected from the chat channel.");
      },

      received(data) {
        const messagesContainer = document.getElementById('messages');
        messagesContainer.insertAdjacentHTML('beforeend', data.message);
      }
    });

    const messageForm = document.getElementById('new_message');
    messageForm.addEventListener('submit', (e) => {
      e.preventDefault();
      const messageContent = document.getElementById('message_content');
      consumer.subscriptions.subscriptions[0].send({
        message: messageContent.value,
        room: roomId
      });
      messageContent.value = '';
    });
  }
});

File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    stream_for "chat_#{params['room']}_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def send_message(data)
    message = Message.create!(content: data['message'], user_id: current_user.id, room_id: data['room'])
    ActionCable.server.broadcast "chat_#{data['room']}_channel", message: render_message(message)
  end

  private

  def render_message(message)
    ApplicationController.renderer.render(partial: 'messages/message', locals: { message: message })
  end
end
